Hw2

CzerPDX opened the pull request:
01-coins.js
- Starts with the value given as `input` and saves it as `currentTotal`
- In order (from largest to smallest denomination) an amount equal from the denomination is subtracted from `currentTotal` until a full amount cannot be taken out.
- For each denomination multiple subtracted a counter is increased for that denomination.
- The output string is built as the subtraction happens. Differentiation between plural and singular words is made.
- Instead of working directly with float, the float is turned into a string. This avoids a rounding issue I was seeing for pennies.

02-reverse.js / 02-reverse.html
- Since I'm not using a form I did the validation manually using `validReverse()`. Validation is performed using a regex pattern.
- Output div starts with a class of `displayNone`, but this will get overwritten with either the `validInput` or `invalidInput` classes once the submit button is clicked.
- Once input is validated the number is fed into reverseNumber it is:
    - Turned into a string
    - Split into separate characters and put into an array using `.split()`
    - Then the array is reversed using `.reverse()`
    - Then joined back into a string using `.join()`

03-form.js / 03-form.html
- Did my best to follow the bootstrap 5.0 guidelines as well as those from the elegant javascript chapter referenced in the assignment.
- Each form element label/input set uses the md-3 bootstrap class for its spacing.
- Each form element has a label for accessibility.
- Correct types were used
- Name and email are required. A basic script from [the bootstrap 5.0 documentation](https://getbootstrap.com/docs/5.0/forms/validation/) was used to prevent form submission unless the data is valid. Data validity is checked using `checkValidity()`.
- There are divs that contain error messages that will only appear if you attempt to submit the form without the required fields or if those fields are invalid.
- Console output is formatted as shown in homework assignment

04-api.js / 04-api.html
- Once the page loads the API call gets made and the data is grabbed from the server
- For each entry the common name of the country is pulled out and saved into a string with `<li>` for each one so it can be inserted and turn into a list in the HTML
- Number is formatted for American numbers with commas.

CzerPDX committed:
added 01-coins.js solution

CzerPDX committed:
Merge branch 'hw2' of github.com:CzerPDX/student-repo-fullstack into hw2

CzerPDX committed:
added solution for 02-reverse.js

CzerPDX committed:
added solution for 03-form.js

smah29 commented on hw2/01-coins.js:
```diff
@@ -1,7 +1,161 @@
+const PENNY_VAL = 0.01;
+
+// Count the multiples of the denomination and return the count
+const countMultiples = (currentTotal, denomination) => {
```
very modular, well commented code, good work

CzerPDX replied:
Thanks!

smah29 commented on hw2/02-reverse.js:
```diff
@@ -1,3 +1,49 @@
 
-// Add your code here
+// Function to reverse string
+function reverseNumber(input) {
```
need to reverse number as number not as string

CzerPDX replied:
Can you give me a little more information on why you feel the reversed input should be a number rather than a string?
The input is taken from the user as a number (and the user is unable to enter anything except a number), however, there is no requirement that I can see in the homework that the reversed output be a number as well. 

Using `toString()` on the input number makes it easier to work with and can provide the required output without issue. Additionally, if the reversed string needed to be used as a real number in some math function, I could easily translate it back from a string into a number using `parseInt()`, though that step seems unnecessary here, as we are simply displaying it to the user.

smah29 replied:
In the HW2, it was mentioned :-
Write a function that reverses the digits in a number.

CzerPDX replied:
Hm, I didn't read it as meaning a datatype, but it's possible that's true and I just misunderstood. Maybe @JeremyIV can confirm one way or the other if the user input needs to remain a number or if it's ok to take in a number but then transform it into a string to reverse it?

JeremyIV replied:
I would consider this to be an implementation detail up to your discretion as a programmer. This problem is about the user interface more than doing some kind of clever numerical manipulations in Javascript. I think treating the number as a string is a very reasonable way to reverse it.

CzerPDX replied:
Thank you!

smah29 commented on hw2/03-form.js:
```diff
@@ -1,3 +1,77 @@
+    }
+
+    // Print message to the console
+    console.log(`========= Form Submission =========
```
not the right way for console.log

CzerPDX replied:
Thank you. Are you able to clarify why you think it is not the right way? Console.log can contain multiline outputs when using the backtick character. The output matches exactly as the professor requested in the assignment. So I am unable to understand why you request a change here.

CzerPDX replied:
Actually, after thinking about it for a little while, I decided it was probably better to build the output as a string and then output the string to `console.log()`. I'm not sure if that's what you meant, but if so, good catch!

smah29 replied:
grouping has to be used see the chapter as mentioned by professor

CzerPDX replied:
Thank you! I completely forgot about `console.group()`. I pushed another commit where I fixed it. Thank you for pointing this out.

smah29 reviewed (dismissed):
Given some comments, Please see

CzerPDX committed:
changed output to console to use console.group()

CzerPDX committed:
edited so buttons are full width and 2 across

CzerPDX committed:
added solution for 04-api.js

gleason9113 reviewed (commented):
Looks good to me.  The main feedback I would offer is that I think that a switch statement might be more readable than an extended if/else block.  Approved!

CzerPDX commented:
> Looks good to me. The main feedback I would offer is that I think that a switch statement might be more readable than an extended if/else block. Approved!

Thanks I will change it to a switch statement tonight or tomorrow!

CzerPDX committed:
removed a bitwise operator that was mistakenly included

CzerPDX committed:
used switch statement in 01-coins.js and added ternary operator

CzerPDX committed:
added try/catch to getMoneyName() in case switch ever gets to default it will throw an error and exit.

CzerPDX committed:
fixed typo

JeremyIV commented:
You have a very thoroughly engineered coins solution! I appreciate the enthusiasm, but I think there are significant opportunities to simplify this code without losing flexibility.

For exercise 3, you should prevent the default behavior on any form submission so that we can see what is logged to the console even when the form is filled. There's a copy-paste error where the email field is logged as "name". 

For exercise 4, it would be better to display an error message on the webpage when the api request fails, not just hide an error message in the console and show an empty page.

Otherwise, your code looks good, nice job!

CzerPDX committed:
03-form changes to form behavior. Disabled default behavior. Changed approach to one that would work for multiple forms on the same page so this has utility for me in the future as a reference. Also fixed typo in console.log output

CzerPDX committed:
04-api.js changes: added error display on page if error occurs loading the country list. Also reused 02-reverse.css and renamed to styles.css so it was more general use.

CzerPDX committed:
removed countMultiples by simplifying. Also stopped using strings and instead worked with cents as integers

CzerPDX committed:
fixed 01-coins.js issues

CzerPDX commented:
> You have a very thoroughly engineered coins solution! I appreciate the enthusiasm, but I think there are significant opportunities to simplify this code without losing flexibility.

Sorry about that! I guess my idea was to try to make fewer anonymous math functions in the main area so it could all be labelled, but that wasn't as helpful as I thought it was.
- Removed `countMultiples()` and replaced with `Math.floor(currentTotal / denominationAmount)` (pseudocode)
- Switched over to treating cents as integers to avoid rounding and remove string conversion. It's possible this actually just made it harder to read because I introduced some `* 100` in here.
- I had trouble imagining how to cut down on the two remaining formatting functions without losing some accuracy (ie comma placement and plurality being correct). Please let me know if you think I should try harder to simplify this though!

> For exercise 3, you should prevent the default behavior on any form submission so that we can see what is logged to the console even when the form is filled. There's a copy-paste error where the email field is logged as "name".
- Fixed copy/paste error in form submission label.
- Changed it so default behavior was disabled. Now no longer refreshes after successful entry.
- Also took a more general approach for validation. This is mostly so I can use it in the future as a more general reference.

> For exercise 4, it would be better to display an error message on the webpage when the api request fails, not just hide an error message in the console and show an empty page.
- Added error to page. Good point!

> Otherwise, your code looks good, nice job!
- Thank you!

CzerPDX committed:
added notes to gitignore

CzerPDX committed:
Merge branch 'main' of github.com:CzerPDX/student-repo-fullstack into hw2
